Python/Product/PythonTools/PythonTools/Project/PytestConfigureInfoBar.cs

```diff
@@ -1,222 +1,298 @@
+        private PyTestInfoBarData InfoBarData { get; set; }
 
+        protected PyTestInfoBar(IServiceProvider site) : base(site) {
 
-            //Simplify syntax TODO
-            var packageManagers = _interpreterOptionsService.GetPackageManagers(_interpreterFactory);
-            var getPackagesTask = new List<Task<IList<PackageSpec>>>();
-            foreach (var packageManager in packageManagers) {
-                getPackagesTask.Add(packageManager.GetInstalledPackagesAsync(CancellationToken.None));
+        }
+
+        protected async Task CheckAsync(PyTestInfoBarData infoBarData) {
+            InfoBarData = infoBarData;
+            if (IsCreated || InfoBarData.IsGloballySuppressed || InfoBarData.InfoBarSuppressed) {
+                return;
             }
 
-            Boolean pytestPackageFound = false;
-            var packages = await Task.WhenAll(getPackagesTask);
-            foreach (var packageSpecs in packages) {
-                var result = packageSpecs.FirstOrDefault(
-                    x => String.Equals(x.Name, nameof(TestFrameworkType.Pytest), StringComparison.OrdinalIgnoreCase
-                ));
+            /*
+            There are 4 possible cases. 
+            #1: Config file is     present, pytest is     installed, pytest is not enabled -> "Enable Pytest"
+            #2: Config file is     present, pytest is not installed, pytest is not enabled -> "Install and enable Pytest"
+            #3: Config file is     present, pytest is not installed, pytest is     enabled -> "Install Pytest"
+            #4: Config file is not present, pytest is not installed, pytest is     enabled -> "Install Pytest"
+             */
+            bool validConfigFile = File.Exists(infoBarData.PyTestConfigFilePath);
+
+            string infoBarMessage;
+            InfoBarHyperlink acceptActionItem;
+
+            if (validConfigFile &&
+                InfoBarData.TestFramework != TestFrameworkType.Pytest
+                && await IsPyTestInstalled()
+            ) {
+                //Case #1. "Enable Pytest"
+                infoBarMessage = Strings.PyTestConfigurationFileFound.FormatUI(InfoBarData.Caption);
+                acceptActionItem = new InfoBarHyperlink(Strings.PyTestEnableInfoBarAction, (Action)EnablePytestAction);
 
-                if (result != null) {
-                    pytestPackageFound = true;
-                    break;
-                }
-            }
+            } else if (validConfigFile &&
+                InfoBarData.TestFramework != TestFrameworkType.Pytest
+                && !(await IsPyTestInstalled())
+            ) {
+                //Case #2. "Install and enable Pytest"
+                infoBarMessage = Strings.PyTestConfigurationFileFound.FormatUI(InfoBarData.Caption);
+                acceptActionItem = new InfoBarHyperlink(
+                    Strings.PyTestInstallAndEnableInfoBarAction,
+                    (Action)(async () => await InstallAndEnablePytestActionAsync())
```

Comment:
that's not going to handle errors.

you need to do

void InstallAndEnablePytest() {
InstallAndEnablePytestAsync().DoNotWait().HandleAllExceptions();
}

Task InstallAndEnablePytestAsync() {
}

Reply:
same for the case below